File: ios/BetterBasicStrategy/Engine/GameEngine.swift
import Foundation

enum GameEngine {

    // MARK: - Deal new round

    static func dealNewRound(
        state: GameState,
        shoe: inout ShoeState,
        rules: RuleSet
    ) -> (state: GameState, reshuffled: Bool) {
        let wasReshuffle = shoe.needsReshuffle
        if wasReshuffle {
            shoe = ShoeState.create(penetrationTrigger: shoe.penetrationTrigger)
        }

        var spots = (0..<rules.numberOfSpots).map { i in
            SpotState(
                id: "spot-\(i)-\(Int(Date().timeIntervalSince1970 * 1000))",
                hands: [HandState.build(from: [])],
                activeHandIndex: 0,
                outcomes: [.lose],
                actions: [[]],
                isComplete: false
            )
        }
        var dealerCards: [Card] = []

        // Two passes: player, dealer, player, dealer-hole
        for _ in 0..<2 {
            for i in 0..<spots.count {
                if let card = shoe.deal() {
                    spots[i].hands[0] = HandState.build(from: spots[i].hands[0].cards + [card])
                }
            }
            if let card = shoe.deal() { dealerCards.append(card) }
        }

        // Mark natural blackjacks
        for i in 0..<spots.count where spots[i].hands[0].isBlackjack {
            spots[i].isComplete = true
            spots[i].outcomes = [.blackjack]
        }

        shoe.save()

        var newState = GameState(
            phase: .playerTurn,
            spots: spots,
            activeSpotIndex: 0,
            dealerCards: dealerCards,
            dealerHoleRevealed: false,
            lastActionWasCorrect: nil,
            lastCoachingEntry: nil
        )

        if spots.allSatisfy(\.isComplete) {
            newState = playDealer(state: newState, shoe: &shoe, rules: rules)
            return (newState, wasReshuffle)
        }

        let firstActive = spots.indices.first(where: { !spots[$0].isComplete }) ?? 0
        newState.activeSpotIndex = firstActive
        return (newState, wasReshuffle)
    }

    // MARK: - Available actions

    static func getAvailableActions(state: GameState, rules: RuleSet) -> Set<DecisionCategory> {
        guard state.phase == .playerTurn else { return [] }
        guard let spot = state.spots[safe: state.activeSpotIndex] else { return [] }
        guard let hand = spot.hands[safe: spot.activeHandIndex]   else { return [] }

        var available: Set<DecisionCategory> = [.hit, .stand]

        if hand.cards.count == 2 {
            let isAfterSplit = spot.hands.count > 1
            if !isAfterSplit || rules.doubleAfterSplit { available.insert(.double) }

            let r1 = hand.cards[0].strategyRank, r2 = hand.cards[1].strategyRank
            if r1 == r2 {
                let isAce = hand.cards[0].rank == 14
                if spot.hands.count == 1 || (spot.hands.count < 4 && (!isAce || rules.resplitAces)) {
                    available.insert(.split)
                }
            }
        }

        if hand.cards.count == 2,
           spot.actions[safe: spot.activeHandIndex]?.isEmpty == true,
           rules.surrenderRule != .none {
            available.insert(.surrender)
        }

        return available
    }

    // MARK: - Perform action

    static func performAction(
        state: GameState,
        shoe: inout ShoeState,
        category: DecisionCategory,
        rules: RuleSet,
        engine: BasicStrategyEngine
    ) -> GameState {
        var spots = state.spots
        var spot  = spots[state.activeSpotIndex]
        let hand  = spot.hands[spot.activeHandIndex]
        let upcard = state.dealerCards[0]

        let available   = getAvailableActions(state: state, rules: rules)
        let canDouble   = available.contains(.double)
        let canSplit    = available.contains(.split)
        let canSurrender = available.contains(.surrender)

        let rawCorrect   = engine.correctAction(handKey: hand.handKey, dealerUpcard: upcard)
        let resolved     = resolveAction(rawCorrect, canDouble: canDouble, canSplit: canSplit, canSurrender: canSurrender)
        let correctCat   = actionCategory(resolved)
        let isCorrect    = correctCat == category
        let entry        = engine.getEntry(handKey: hand.handKey, dealerUpcard: upcard)

        spot.actions[spot.activeHandIndex].append(ActionRecord(
            action: category,
            wasCorrect: isCorrect,
            correctAction: resolved,
            explanation: entry?.explanation ?? ""
        ))

        var newState = state
        newState.spots = spots
        newState.lastActionWasCorrect = isCorrect
        newState.lastCoachingEntry    = isCorrect ? nil : entry

        switch category {
        case .hit:
            if let card = shoe.deal() {
                spot.hands[spot.activeHandIndex] = HandState.build(from: hand.cards + [card])
                spots[state.activeSpotIndex] = spot
                newState.spots = spots
                if spot.hands[spot.activeHandIndex].total >= 21 {
                    newState = advanceHand(state: newState, shoe: &shoe, rules: rules)
                }
            } else {
                spots[state.activeSpotIndex] = spot; newState.spots = spots
            }

        case .stand:
            spots[state.activeSpotIndex] = spot; newState.spots = spots
            newState = advanceHand(state: newState, shoe: &shoe, rules: rules)

        case .double:
            if let card = shoe.deal() {
                spot.hands[spot.activeHandIndex] = HandState.build(from: hand.cards + [card])
            }
            spots[state.activeSpotIndex] = spot; newState.spots = spots
            newState = advanceHand(state: newState, shoe: &shoe, rules: rules)

        case .split:
            let splitCard = hand.cards[1]
            let fill1 = shoe.deal()
            let fill2 = shoe.deal()
            spot.hands[spot.activeHandIndex] = HandState.build(from: [hand.cards[0]] + (fill1.map { [$0] } ?? []))
            let newHand = HandState.build(from: [splitCard] + (fill2.map { [$0] } ?? []))
            spot.hands.insert(newHand, at: spot.activeHandIndex + 1)
            spot.actions.insert([], at: spot.activeHandIndex + 1)
            spot.outcomes.append(.lose)
            spots[state.activeSpotIndex] = spot; newState.spots = spots
            // Split aces get only one card each unless resplitAces
            if hand.cards[0].rank == 14 && !rules.resplitAces {
                newState = advanceHand(state: newState, shoe: &shoe, rules: rules)
            }

        case .surrender:
            spot.outcomes = [.surrender]; spot.isComplete = true
            spots[state.activeSpotIndex] = spot; newState.spots = spots
            newState = advanceIfNeeded(state: newState, shoe: &shoe, rules: rules)
        }

        shoe.save()
        return newState
    }

    // MARK: - Advance helpers

    private static func advanceHand(state: GameState, shoe: inout ShoeState, rules: RuleSet) -> GameState {
        var newState = state
        var spot = newState.spots[newState.activeSpotIndex]
        let next = spot.activeHandIndex + 1
        if next < spot.hands.count {
            spot.activeHandIndex = next
            newState.spots[newState.activeSpotIndex] = spot
        } else {
            spot.isComplete = true
            newState.spots[newState.activeSpotIndex] = spot
            newState = advanceIfNeeded(state: newState, shoe: &shoe, rules: rules)
        }
        return newState
    }

    private static func advanceIfNeeded(state: GameState, shoe: inout ShoeState, rules: RuleSet) -> GameState {
        let next = state.spots.indices.first(where: { $0 > state.activeSpotIndex && !state.spots[$0].isComplete })
        if let next = next {
            var s = state; s.activeSpotIndex = next; return s
        }
        if state.spots.allSatisfy(\.isComplete) {
            return playDealer(state: state, shoe: &shoe, rules: rules)
        }
        return state
    }

    // MARK: - Dealer play

    private static func playDealer(state: GameState, shoe: inout ShoeState, rules: RuleSet) -> GameState {
        var dealerCards = state.dealerCards
        while shouldDealerHit(cards: dealerCards, rules: rules) {
            guard let card = shoe.deal() else { break }
            dealerCards.append(card)
        }
        var resolved = resolveOutcomes(state: GameState(
            phase: .dealerTurn, spots: state.spots, activeSpotIndex: state.activeSpotIndex,
            dealerCards: dealerCards, dealerHoleRevealed: true,
            lastActionWasCorrect: state.lastActionWasCorrect, lastCoachingEntry: state.lastCoachingEntry
        ))
        resolved.phase = .roundOver
        shoe.save()
        return resolved
    }

    private static func shouldDealerHit(cards: [Card], rules: RuleSet) -> Bool {
        let (total, isSoft) = calcTotal(cards)
        if total < 17 { return true }
        if total == 17 && isSoft && rules.dealerRule == .H17 { return true }
        return false
    }

    private static func resolveOutcomes(state: GameState) -> GameState {
        let (dealerTotal, _) = calcTotal(state.dealerCards)
        let dealerBust = dealerTotal > 21
        let spots = state.spots.map { spot -> SpotState in
            if spot.outcomes.first == .surrender { return spot }
            let outcomes: [Outcome] = spot.hands.enumerated().map { (i, hand) in
                if spot.outcomes[safe: i] == .blackjack { return .blackjack }
                if hand.total > 21  { return .bust }
                if dealerBust       { return .win }
                if hand.total > dealerTotal { return .win }
                if hand.total == dealerTotal { return .push }
                return .lose
            }
            return SpotState(id: spot.id, hands: spot.hands, activeHandIndex: spot.activeHandIndex,
                             outcomes: outcomes, actions: spot.actions, isComplete: true)
        }
        var s = state; s.spots = spots; s.dealerHoleRevealed = true; return s
    }
}

// MARK: - Safe subscript

extension Array {
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

File: expo-app/src/engine/gameEngine.ts
import type { ActionRecord, Card, DecisionCategory, GameState, Outcome, RuleSet, SpotState } from '../types'
import { makeCard } from './card'
import { buildHandState, calcTotal } from './hand'
import { dealCard, needsReshuffle, createShoe } from './shoe'
import type { ShoeState } from './shoe'
import { BasicStrategyEngine, resolveAction, actionCategory } from './basicStrategyEngine'
import { strategyRank } from './card'

function makeSpot(id: string): SpotState {
  return { id, hands: [buildHandState([])], activeHandIndex: 0, outcomes: ['lose'], actions: [[]], isComplete: false }
}

export function initialGameState(): GameState {
  return { phase: 'idle', spots: [], activeSpotIndex: 0, dealerCards: [], dealerHoleRevealed: false, lastActionWasCorrect: null, lastCoachingEntry: null }
}

// ─── Deal ────────────────────────────────────────────────────────────────────

export function dealNewRound(
  state: GameState, shoe: ShoeState, rules: RuleSet
): { state: GameState; shoe: ShoeState; reshuffled: boolean } {
  const wasReshuffle = needsReshuffle(shoe)
  let s = wasReshuffle ? createShoe(shoe.penetrationTrigger) : shoe

  const spots: SpotState[] = Array.from({ length: rules.numberOfSpots }, (_, i) =>
    makeSpot(`spot-${i}-${Date.now()}`)
  )
  const dealerCards: Card[] = []

  for (let pass = 0; pass < 2; pass++) {
    for (let i = 0; i < spots.length; i++) {
      const result = dealCard(s)
      if (!result) continue
      s = result.shoe
      spots[i].hands[0] = buildHandState([...spots[i].hands[0].cards, result.card])
    }
    const result = dealCard(s)
    if (result) { dealerCards.push(result.card); s = result.shoe }
  }

  for (const spot of spots) {
    if (spot.hands[0].isBlackjack) { spot.isComplete = true; spot.outcomes = ['blackjack'] }
  }

  const newState: GameState = { phase: 'playerTurn', spots, activeSpotIndex: 0, dealerCards, dealerHoleRevealed: false, lastActionWasCorrect: null, lastCoachingEntry: null }

  if (spots.every(sp => sp.isComplete)) {
    const r = playDealer({ ...newState }, s, rules)
    return { state: r.state, shoe: r.shoe, reshuffled: wasReshuffle }
  }

  const first = spots.findIndex(sp => !sp.isComplete)
  newState.activeSpotIndex = first >= 0 ? first : 0
  return { state: newState, shoe: s, reshuffled: wasReshuffle }
}

// ─── Player actions ───────────────────────────────────────────────────────────

export function getAvailableActions(state: GameState, rules: RuleSet): Set<DecisionCategory> {
  if (state.phase !== 'playerTurn') return new Set()
  const spot = state.spots[state.activeSpotIndex]
  if (!spot) return new Set()
  const hand = spot.hands[spot.activeHandIndex]
  if (!hand) return new Set()

  const available: Set<DecisionCategory> = new Set(['hit', 'stand'])

  if (hand.cards.length === 2) {
    const isAfterSplit = spot.hands.length > 1
    if (!isAfterSplit || rules.doubleAfterSplit) available.add('double')

    const r1 = strategyRank(hand.cards[0].rank)
    const r2 = strategyRank(hand.cards[1].rank)
    if (r1 === r2) {
      const handsCount = spot.hands.length
      const isAce = hand.cards[0].rank === 14
      if (handsCount < 4 && (!isAce || rules.resplitAces)) available.add('split')
      if (handsCount === 1) available.add('split')
    }
  }

  if (hand.cards.length === 2 && spot.actions[spot.activeHandIndex].length === 0 && rules.surrenderRule !== 'none') {
    available.add('surrender')
  }

  return available
}

export function performAction(
  state: GameState, shoe: ShoeState, category: DecisionCategory, rules: RuleSet, engine: BasicStrategyEngine
): { state: GameState; shoe: ShoeState } {
  let s = shoe
  const spots = state.spots.map(sp => ({ ...sp, hands: [...sp.hands], actions: sp.actions.map(a => [...a]), outcomes: [...sp.outcomes] }))
  const spot = spots[state.activeSpotIndex]
  const hand = spot.hands[spot.activeHandIndex]
  const dealerUpcard = state.dealerCards[0]

  const available   = getAvailableActions(state, rules)
  const canDouble   = available.has('double')
  const canSplit    = available.has('split')
  const canSurrender = available.has('surrender')

  const rawCorrect    = engine.correctAction(hand.handKey, dealerUpcard)
  const resolved      = resolveAction(rawCorrect, canDouble, canSplit, canSurrender)
  const correctCat    = actionCategory(resolved)
  const isCorrect     = correctCat === category
  const entry         = engine.getEntry(hand.handKey, dealerUpcard)

  const record: ActionRecord = { action: category, wasCorrect: isCorrect, correctAction: resolved, explanation: entry?.explanation ?? '' }
  spot.actions[spot.activeHandIndex] = [...spot.actions[spot.activeHandIndex], record]

  let newState: GameState = { ...state, spots, lastActionWasCorrect: isCorrect, lastCoachingEntry: isCorrect ? null : (entry ?? null) }

  switch (category) {
    case 'hit': {
      const result = dealCard(s)
      if (result) {
        s = result.shoe
        spot.hands[spot.activeHandIndex] = buildHandState([...hand.cards, result.card])
        if (spot.hands[spot.activeHandIndex].total >= 21) {
          return { state: advanceHand({ ...newState, spots }, s, rules), shoe: s }
        }
      }
      break
    }
    case 'stand':
      return { state: advanceHand({ ...newState, spots }, s, rules), shoe: s }
    case 'double': {
      const result = dealCard(s)
      if (result) { s = result.shoe; spot.hands[spot.activeHandIndex] = buildHandState([...hand.cards, result.card]) }
      return { state: advanceHand({ ...newState, spots }, s, rules), shoe: s }
    }
    case 'split': {
      const splitCard = hand.cards[1]
      const fill1 = dealCard(s); if (fill1) s = fill1.shoe
      const fill2 = dealCard(s); if (fill2) s = fill2.shoe
      spot.hands[spot.activeHandIndex] = buildHandState([hand.cards[0], ...(fill1 ? [fill1.card] : [])])
      const newHand = buildHandState([splitCard, ...(fill2 ? [fill2.card] : [])])
      spot.hands = [...spot.hands.slice(0, spot.activeHandIndex + 1), newHand, ...spot.hands.slice(spot.activeHandIndex + 1)]
      spot.actions = [...spot.actions.slice(0, spot.activeHandIndex + 1), [], ...spot.actions.slice(spot.activeHandIndex + 1)]
      spot.outcomes = [...spot.outcomes, 'lose']
      if (hand.cards[0].rank === 14 && !rules.resplitAces) {
        return { state: advanceHand({ ...newState, spots }, s, rules), shoe: s }
      }
      break
    }
    case 'surrender':
      spot.outcomes = ['surrender']; spot.isComplete = true
      return { state: advanceIfNeeded({ ...newState, spots }, s, rules), shoe: s }
  }

  return { state: { ...newState, spots }, shoe: s }
}

// ─── Internal helpers ─────────────────────────────────────────────────────────

function advanceHand(state: GameState, shoe: ShoeState, rules: RuleSet): GameState {
  const spots = [...state.spots]
  const spot = { ...spots[state.activeSpotIndex] }
  const next = spot.activeHandIndex + 1
  if (next < spot.hands.length) {
    spots[state.activeSpotIndex] = { ...spot, activeHandIndex: next }
    return { ...state, spots }
  }
  spots[state.activeSpotIndex] = { ...spot, isComplete: true }
  return advanceIfNeeded({ ...state, spots }, shoe, rules)
}

function advanceIfNeeded(state: GameState, shoe: ShoeState, rules: RuleSet): GameState {
  const next = state.spots.findIndex((sp, i) => !sp.isComplete && i > state.activeSpotIndex)
  if (next >= 0) return { ...state, activeSpotIndex: next }
  if (state.spots.every(sp => sp.isComplete)) return playDealer(state, shoe, rules).state
  return state
}

function playDealer(state: GameState, shoe: ShoeState, rules: RuleSet): { state: GameState; shoe: ShoeState } {
  let s = shoe
  let dealerCards = [...state.dealerCards]
  while (shouldDealerHit(dealerCards, rules)) {
    const result = dealCard(s)
    if (!result) break
    s = result.shoe; dealerCards = [...dealerCards, result.card]
  }
  const resolved = resolveOutcomes({ ...state, dealerCards, dealerHoleRevealed: true, phase: 'dealerTurn' })
  return { state: { ...resolved, phase: 'roundOver' }, shoe: s }
}

function shouldDealerHit(cards: Card[], rules: RuleSet): boolean {
  const { total, isSoft } = calcTotal(cards)
  if (total < 17) return true
  if (total === 17 && isSoft && rules.dealerRule === 'H17') return true
  return false
}

function resolveOutcomes(state: GameState): GameState {
  const { total: dealerTotal } = calcTotal(state.dealerCards)
  const dealerBust = dealerTotal > 21
  const spots = state.spots.map(spot => {
    if (spot.outcomes[0] === 'surrender') return spot
    const outcomes: Outcome[] = spot.hands.map((hand, i) => {
      if (spot.outcomes[i] === 'blackjack') return 'blackjack'
      if (hand.total > 21) return 'bust'
      if (dealerBust) return 'win'
      if (hand.total > dealerTotal) return 'win'
      if (hand.total === dealerTotal) return 'push'
      return 'lose'
    })
    return { ...spot, outcomes, isComplete: true }
  })
  return { ...state, spots, dealerHoleRevealed: true }
}
